source/core/index.ts

```diff
@@ -0,0 +1,1472 @@
+				this._beforeError(new UploadError(error, options, this));
+			});
+
+			options.body.once('end', () => {
+				delete options.body;
+			});
+		} else {
+			this._unlockWrite();
+
+			if (!is.undefined(options.body)) {
+				this._writeRequest(options.body, null as unknown as string, () => {});
+				currentRequest.end();
+
+				this._lockWrite();
+			} else if (this._cannotHaveBody || this._noPipe) {
+				currentRequest.end();
+
+				this._lockWrite();
+			}
+		}
+
+		this.emit('request', request);
+	}
+
+	async _makeRequest(): Promise<void> {
+		const {options} = this;
+		const {url, headers, request, agent, timeout} = options;
+
+		for (const key in headers) {
+			if (is.undefined(headers[key])) {
+				// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+				delete headers[key];
+			} else if (is.null_(headers[key])) {
+				throw new TypeError(`Use \`undefined\` instead of \`null\` to delete the \`${key}\` header`);
+			}
+		}
+
+		if (options.decompress && is.undefined(headers['accept-encoding'])) {
+			headers['accept-encoding'] = supportsBrotli ? 'gzip, deflate, br' : 'gzip, deflate';
+		}
+
+		// Set cookies
+		if (options.cookieJar) {
+			const cookieString: string = await options.cookieJar.getCookieString(options.url.toString());
+
+			if (is.nonEmptyString(cookieString)) {
+				options.headers.cookie = cookieString;
+			}
+		}
+
+		for (const hook of options.hooks.beforeRequest) {
+			// eslint-disable-next-line no-await-in-loop
+			const result = await hook(options);
+
+			if (!is.undefined(result)) {
+				// @ts-ignore Skip the type mismatch to support abstract responses
+				options.request = () => result;
+				break;
+			}
+		}
+
+		if (options.dnsCache && !('lookup' in options)) {
+			options.lookup = options.dnsCache.lookup;
+		}
+
+		// UNIX sockets
+		if (url.hostname === 'unix') {
+			const matches = /(?<socketPath>.+?):(?<path>.+)/.exec(`${url.pathname}${url.search}`);
+
+			if (matches?.groups) {
+				const {socketPath, path} = matches.groups;
+
+				Object.assign(options, {
+					socketPath,
+					path,
+					host: ''
+				});
+			}
+		}
+
+		const isHttps = url.protocol === 'https:';
+
+		let fallbackFn: HttpRequestFunction;
+		if (options.http2) {
+			fallbackFn = http2wrapper.auto;
+		} else {
+			fallbackFn = isHttps ? https.request : http.request;
+		}
+
+		const realFn = options.request ?? fallbackFn;
+		const fn = options.cacheableRequest ? cacheFn : realFn;
+
+		if (agent && !options.http2) {
+			(options as unknown as RequestOptions).agent = agent[isHttps ? 'https' : 'http'];
+		}
+
+		options[kRequest] = realFn as HttpRequestFunction;
+		delete options.request;
+		delete options.timeout;
+
+		let requestOrResponse: ReturnType<RequestFunction>;
+
+		try {
+			requestOrResponse = await fn(url, options as unknown as RequestOptions);
+
+			if (is.undefined(requestOrResponse)) {
+				requestOrResponse = fallbackFn(url, options as unknown as RequestOptions);
```

Comment:
Future behavior: if `options.request` returns `undefined` (e.g. cache mechanism: no cache found), fallback to native HTTP requests.